api/aws/provider.go

```diff
@@ -0,0 +1,99 @@
+package aws
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"github.com/riotgames/key-conjurer/api/core"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/sts"
+)
+
+type ErrRoleNotFound struct{ Name string }
+
+func (e ErrRoleNotFound) Error() string {
+	return fmt.Sprintf("role %s was not found or you do not have access to it", e.Name)
+}
+
+type Provider struct {
+	stsClient *sts.STS
+}
+
+func NewProvider(region string) (*Provider, error) {
+	config, err := session.NewSession(&aws.Config{Region: aws.String(region)})
+	if err != nil {
+		return nil, err
+	}
+
+	return &Provider{stsClient: sts.New(config)}, nil
+}
+
+type STSTokenResponse struct {
+	AccessKeyID     *string `json:"accessKeyId"`
+	SecretAccessKey *string `json:"secretAccessKey"`
+	SessionToken    *string `json:"sessionToken"`
+	Expiration      string  `json:"expiration"`
+}
+
+func getRole(roleName string, response *core.SAMLResponse) (string, string, error) {
+	if response == nil {
+		return "", "", errors.New("Unable to get SAML assertion")
+	}
+
+	roles := strings.Split(response.GetAttribute("https://aws.amazon.com/SAML/Attributes/Role"), ",")
+	if roleName == "" {
+		// This is for legacy support.
+		// Legacy clients would always retrieve the first two ARNs in the list, which would be
+		//   arn:aws:iam::[account-id]:role/[onelogin_role]
+		//   arn:aws:iam::[account-id]:saml-provider/[saml-provider]
+		// If we get weird breakages with Key Conjurer when it's deployed alongside legacy clients, this is almost certainly a culprit!
+		return roles[1], roles[0], nil
+	}
+
+	var roleARN string
+	for _, arn := range roles[1:] {
```

Comment:
Warning: This is broken for users with multiple roles in the SAML assertion. There is a follow-up PR to this one to resolve this.